compositor_common/src/scene.rs

```diff
@@ -1,41 +1,83 @@
 use serde::{Deserialize, Serialize};
-use std::{any::Any, collections::HashMap, sync::Arc};
+use std::{collections::HashMap, sync::Arc};
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
-pub struct VideoId(usize);
-
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct TransformationRegistryKey(pub String);
+use crate::transformation::TransformationRegistryKey;
 
 #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
 pub struct Resolution {
     pub width: usize,
     pub height: usize,
 }
 
-#[derive(Debug, Clone, Hash, PartialEq, Eq)]
-pub struct NodeId(pub String);
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct NodeId(pub Arc<str>);
 
-#[derive(Debug)]
-pub enum Node {
-    Video {
-        id: VideoId,
-        resolution: Resolution,
-    },
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct InputId(pub NodeId);
 
-    Image {
-        data: Vec<u8>,
-        resolution: Resolution,
-    },
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct OutputId(pub NodeId);
+
+impl From<NodeId> for InputId {
+    fn from(value: NodeId) -> Self {
+        Self(value)
+    }
+}
+
+impl From<NodeId> for OutputId {
+    fn from(value: NodeId) -> Self {
+        Self(value)
+    }
+}
+
+/// SceneSpec represents configuration that can be used to create new Scene
+/// or update an existing one.
+#[derive(Serialize, Deserialize)]
+pub struct SceneSpec {
+    pub inputs: Vec<InputSpec>,
+    pub transforms: Vec<TransformNodeSpec>,
+    pub outputs: Vec<OutputSpec>,
+}
 
-    Transformer {
-        registry_key: TransformationRegistryKey,
-        inputs: HashMap<NodeId, Arc<Node>>,
-        resolution: Resolution,
-        params: Box<dyn Any>,
+#[derive(Serialize, Deserialize)]
+pub struct InputSpec {
+    pub input_id: InputId,
+    pub resolution: Resolution,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct OutputSpec {
+    pub output_id: OutputId,
+    pub input_pad: NodeId,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct TransformNodeSpec {
+    pub node_id: NodeId,
+    pub input_pads: Vec<NodeId>,
+    pub resolution: Resolution,
+
+    #[serde(flatten)]
+    pub transform_params: TransformParams,
+}
+
+// TODO: tmp clone
+#[derive(Serialize, Deserialize, Clone)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum TransformParams {
+    WebRenderer {
+        renderer_id: TransformationRegistryKey,
+    },
+    Shader {
+        shader_id: TransformationRegistryKey,
+        shader_params: HashMap<String, ShaderParams>,
     },
```

Comment:
We need to allow to specify the order of `shader_params`, to pass them correctly to shader uniform. `HashMap` is an inherently unordered data structure, so I think we should use `List` here.

Moreover, `ShaderParams` require casting all passed values to bits by a user and sending it in JSON. I think that we want to do that in a more "human" way like:
```
[
    "f32": 3.1415,
    "u32": 2137
]
```

cc @jerzywilczek

Reply:
As mentioned in the previous comment this is just a placeholder. Also just using a List like you described wouldn't be enough, but that is not really a discussion for this PR.